agent/app/agent.go

```diff
@@ -0,0 +1,404 @@
+	}, nil
+
+}
+
+func (agent *ecsAgent) printVersion() int {
+	version.PrintVersion(agent.dockerClient)
+	return exitcodes.ExitSuccess
+}
+
+func (agent *ecsAgent) start() int {
+	sighandlers.StartDebugHandler()
+
+	containerChangeEventStream := eventstream.NewEventStream(containerChangeEventStreamName, agent.ctx)
+	credentialsManager := credentials.NewManager()
+	state := dockerstate.NewTaskEngineState()
+	imageManager := engine.NewImageManager(agent.cfg, agent.dockerClient, state)
+
+	return agent.doStart(containerChangeEventStream, credentialsManager, state, imageManager)
+}
+
+func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStream,
+	credentialsManager credentials.Manager,
+	state dockerstate.TaskEngineState,
+	imageManager engine.ImageManager) int {
+
+	taskEngine, currentEC2InstanceID, err := agent.newTaskEngine(containerChangeEventStream,
+		credentialsManager, state, imageManager)
+	if err != nil {
+		return exitcodes.ExitTerminal
+	}
+
+	stateManager, err := agent.newStateManager(taskEngine,
+		&agent.cfg.Cluster, &agent.containerInstanceArn, &currentEC2InstanceID)
+	if err != nil {
+		log.Criticalf("Error creating state manager: %v", err)
+		return exitcodes.ExitTerminal
+	}
+
+	client := ecsclient.NewECSClient(agent.credentialProvider, agent.cfg, agent.ec2MetadataClient)
+	err = agent.registerContainerInstance(taskEngine, stateManager, client)
+	if err != nil {
+		if isNonTerminal(err) {
+			return exitcodes.ExitError
+		}
+		return exitcodes.ExitTerminal
+	}
+
+	// Begin listening to the docker daemon and saving changes
+	taskEngine.SetSaver(stateManager)
+	imageManager.SetSaver(stateManager)
+	taskEngine.MustInit()
+
+	deregisterInstanceEventStream := eventstream.NewEventStream(
+		deregisterContainerInstanceEventStreamName, agent.ctx)
+	deregisterInstanceEventStream.StartListening()
+
+	taskHandler := eventhandler.NewTaskHandler()
+
+	agent.startAsyncRoutines(containerChangeEventStream, credentialsManager, imageManager,
+		taskEngine, stateManager, deregisterInstanceEventStream, client, taskHandler)
+	return agent.startACSSession(credentialsManager, taskEngine, stateManager,
+		deregisterInstanceEventStream, client, taskHandler)
+}
+
+func (agent *ecsAgent) newTaskEngine(containerChangeEventStream *eventstream.EventStream,
+	credentialsManager credentials.Manager,
+	state dockerstate.TaskEngineState,
+	imageManager engine.ImageManager) (engine.TaskEngine, string, error) {
+
+	containerChangeEventStream.StartListening()
+
+	if !agent.cfg.Checkpoint {
+		log.Info("Checkpointing not enabled; a new container instance will be created each time the agent is run")
+		return engine.NewTaskEngine(agent.cfg, agent.dockerClient, credentialsManager,
+			containerChangeEventStream, imageManager, state), "", nil
+	}
+
+	// We try to set these values by loading the existing state file first
+	var previousCluster, previousEC2InstanceID, previousContainerInstanceArn string
```

Comment:
The loaded (previous) ec2 instance id can be different from the current id if launching from an AMI that was created by creating a snapshot of an instance that was registered with ECS.